leader-election-impl/src/main/java/com/palantir/paxos/MigrationPaxosStateLog.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.paxos;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.palantir.common.base.Throwables;
+import com.palantir.common.persist.Persistable;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+
+public final class MigrationPaxosStateLog<V extends Persistable & Versionable> implements PaxosStateLog<V> {
+    private static final Logger log = LoggerFactory.getLogger(MigrationPaxosStateLog.class);
+
+    private final ReadWriteLock lock = new ReentrantReadWriteLock(false);
+    private final PaxosStateLog<V> legacyLog;
+    private final PaxosStateLog<V> currentLog;
+    private final Persistable.Hydrator<V> hydrator;
+
+    private MigrationPaxosStateLog(Settings<V> settings) {
+        this.legacyLog = settings.sourceOfTruth();
+        this.currentLog = settings.secondary();
+        this.hydrator = settings.hydrator();
+    }
+
+    public static <V extends Persistable & Versionable> PaxosStateLog<V> create(Settings<V> settings) {
+        seedCurrentFromLegacy(settings);
```

Comment:
How will you handle remembering to run this only once?

Also, how would we make sure someone doesn’t downgrade from when we start writing only to new store to code that only reads from old store

Reply:
I think the current approach _is_ intentionally to do it everytime. This is probably mitigated by us being non-leader when starting up, generally speaking, so not terrible. Though it could be O(minutes) at our big deployments which seems less than ideal. 

We should think if there's a way to only do this for values that weren't already copied (possibly bounded by something in the learner sequence). If I'm not mistaken overwrites _can_ happen at previous sequence numbers, so we do need to be careful.

Downgrade is probably protected via schema versions in internal framework.